### PR TITLE
Fix TrayUIInstance ptr NULL check to allow it to run on 23H2 Insider

### PR DESCRIPTION
**Pull request by NyaMisty**

### CODE REVIEW

**Comment by NyaMisty** [comment]
I can also help regularly check the compatibility with Dev Insider versions

**Comment by DavidBerdik** [comment]
Would you be willing to test if this PR fixes https://github.com/valinet/ExplorerPatcher/issues/1116?

**Comment by NyaMisty** [comment]
@DavidBerdik Nope, this PR does not fixes anything. 
In my environment, the Windows 10 taskbar from EP will directly crash the explorer.
This PR is only mean to make EP able to start in newer insider without taskbar tweak to allow other tweaking work.

File: ExplorerPatcher/dllmain.c [hunk 1]
```diff
@@ -2491,14 +2491,16 @@ INT64 Shell_TrayWndSubclassProc(
                 HWND hShellTray_Wnd = FindWindowExW(NULL, NULL, L"Shell_TrayWnd", NULL);
                 INT64* CTrayInstance = (BYTE*)(GetWindowLongPtrW(hShellTray_Wnd, 0)); // -> CTray
                 const unsigned int TRAYUI_OFFSET_IN_CTRAY = 110;
-                uintptr_t TrayUIInstance = *((INT64*)CTrayInstance + TRAYUI_OFFSET_IN_CTRAY) + 8;
+                uintptr_t TrayUIInstance = *((INT64*)CTrayInstance + TRAYUI_OFFSET_IN_CTRAY);
```
Comment (valinet) [inline ExplorerPatcher/dllmain.c:2494]: So TrayUI is now the first thing in the struct? Where have you got that from, can you reference some disassembly?

Comment (valinet) [inline ExplorerPatcher/dllmain.c:2494]: So you're saying the pointer to the instance of the `TrayUI` class now lives at offset 0 in the `CTray` class?

File: ExplorerPatcher/dllmain.c [hunk 2]
```diff
@@ -2491,14 +2491,16 @@ INT64 Shell_TrayWndSubclassProc(
                 HWND hShellTray_Wnd = FindWindowExW(NULL, NULL, L"Shell_TrayWnd", NULL);
                 INT64* CTrayInstance = (BYTE*)(GetWindowLongPtrW(hShellTray_Wnd, 0)); // -> CTray
                 const unsigned int TRAYUI_OFFSET_IN_CTRAY = 110;
-                uintptr_t TrayUIInstance = *((INT64*)CTrayInstance + TRAYUI_OFFSET_IN_CTRAY) + 8;
+                uintptr_t TrayUIInstance = *((INT64*)CTrayInstance + TRAYUI_OFFSET_IN_CTRAY);
                 if (TrayUIInstance)
                 {
+                    TrayUIInstance += 8;
```
Comment (valinet) [inline ExplorerPatcher/dllmain.c:2497]: But no, still on the first position. Then what does that `if` check above actually check?

File: ExplorerPatcher/dllmain.c [hunk 3]
```diff
@@ -2491,14 +2491,16 @@ INT64 Shell_TrayWndSubclassProc(
...
                 const unsigned int TRAYUI_OFFSET_IN_CTRAY = 110;
-                uintptr_t TrayUIInstance = *((INT64*)CTrayInstance + TRAYUI_OFFSET_IN_CTRAY) + 8;
+                uintptr_t TrayUIInstance = *((INT64*)CTrayInstance + TRAYUI_OFFSET_IN_CTRAY);
                 if (TrayUIInstance)
                 {
+                    TrayUIInstance += 8;
                     int offset = 656;
                     if (IsWindows11Version22H2OrHigher()) offset = 640;
-                    if ((*(unsigned __int8(__fastcall**)(INT64))(**(INT64**)(TrayUIInstance + offset) + 104i64))(*(INT64*)(TrayUIInstance + offset)))
+                    if (!**(INT64**)(TrayUIInstance + offset)
```
Comment (valinet) [inline ExplorerPatcher/dllmain.c:2500]: So we actually do not know where `TrayUI` is anymore, so we ignore it altogether and delete the `Undo` item regardless?

File: ExplorerPatcher/dllmain.c [hunk 4]
```diff
@@ -2491,14 +2491,16 @@ INT64 Shell_TrayWndSubclassProc(
...
                 {
+                    TrayUIInstance += 8;
                     int offset = 656;
                     if (IsWindows11Version22H2OrHigher()) offset = 640;
-                    if ((*(unsigned __int8(__fastcall**)(INT64))(**(INT64**)(TrayUIInstance + offset) + 104i64))(*(INT64*)(TrayUIInstance + offset)))
+                    if (!**(INT64**)(TrayUIInstance + offset) 
+                        || (*(unsigned __int8(__fastcall**)(INT64))(**(INT64**)(TrayUIInstance + offset) + 104i64))(*(INT64*)(TrayUIInstance + offset)) )
                     {
-                        DeleteMenu(hSubMenu, 0x1A0u, 0);
+                        DeleteMenu(hSubMenu, 416, MF_BYCOMMAND); // Undo
```
Comment (valinet) [inline ExplorerPatcher/dllmain.c:2503]: Aligning with the documentation is fine, my original code probably came from the disassembly.

**Comment by valinet** [comment]
Yeah, so I'd rather have this fixed properly, by determining the offset for those 2 functions in the `TrayUI` struct, or the location where the pointer to `TrayUI` is now located in `CTray`, or what happened with it at all...

**Comment by pieterza** [comment]
> Yeah, so I'd rather have this fixed properly, by determining the offset for those 2 functions in the `TrayUI` struct, or the location where the pointer to `TrayUI` is now located in `CTray`, or what happened with it at all...

Maybe do that then since latest is crashing now after updating windows.  Had to uninstall to be able to work.

**Comment by paunovic** [comment]
Any alternatives to this project? Seems dead.

**Pull request closed without merging**